full-service/src/service/signed_contingent_input.rs

```diff
@@ -0,0 +1,572 @@
+    ///| `sci_proto`        | The signed contingent input profobuf (hex-encoded)               |                                                                                   |
+    ///
+    fn validate_proof_of_reserve_sci(
+        &self,
+        sci_proto: &str,
+    ) -> Result<ValidateProofOfReserveSciResult, SignedContingentInputServiceError>;
+}
+
+impl<T, FPR> SignedContingentInputService for WalletService<T, FPR>
+where
+    T: BlockchainConnection + UserTxConnection + 'static,
+    FPR: FogPubkeyResolver + Send + Sync + 'static,
+{
+    fn validate_proof_of_reserve_sci(
+        &self,
+        sci_proto: &str,
+    ) -> Result<ValidateProofOfReserveSciResult, SignedContingentInputServiceError> {
+        let sci: SignedContingentInput = mc_util_serial::decode(&hex::decode(sci_proto)?)?;
+
+        if let Err(err) = sci.validate() {
+            return Ok(ValidateProofOfReserveSciResult::InvalidSci {
+                error: err.to_string(),
+            });
+        }
+
+        if sci.tx_in.ring.len() != 1 {
+            return Ok(ValidateProofOfReserveSciResult::NotProofOfReserveSci {
+                error: "Proof of reserve SCI must have exactly one ring member".to_string(),
+            });
+        }
+
+        let ring_txo = &sci.tx_in.ring[0];
+        let tx_out_public_key = hex::encode(ring_txo.public_key);
+        let txo_index = match self
+            .ledger_db
+            .get_tx_out_index_by_public_key(&ring_txo.public_key)
+        {
+            Ok(txo_index) => txo_index,
+            Err(LedgerError::NotFound) => {
+                return Ok(ValidateProofOfReserveSciResult::TxOutNotFoundInLedger {
+                    tx_out_public_key,
+                });
+            }
+            Err(e) => return Err(SignedContingentInputServiceError::LedgerError(e)),
+        };
+        let ledger_txo = match self.ledger_db.get_tx_out_by_index(txo_index) {
+            Ok(txo) => txo,
+            Err(LedgerError::NotFound) => {
+                return Ok(ValidateProofOfReserveSciResult::TxOutNotFoundInLedger {
+                    tx_out_public_key,
+                });
+            }
+            Err(e) => return Err(SignedContingentInputServiceError::LedgerError(e)),
+        };
+        if ring_txo != &ledger_txo {
+            return Ok(ValidateProofOfReserveSciResult::TxOutMismatch { tx_out_public_key });
+        }
+
+        if sci.required_output_amounts.len() != 1 {
+            return Ok(ValidateProofOfReserveSciResult::NotProofOfReserveSci {
+                error: "Proof of reserve SCI must have exactly one required output".to_string(),
+            });
+        }
+
+        let required_output_amount = &sci.required_output_amounts[0];
+        if required_output_amount.token_id != sci.pseudo_output_amount.token_id {
+            return Ok(ValidateProofOfReserveSciResult::NotProofOfReserveSci {
+                error: "Proof of reserve SCI must have the same token id as the pseudo-output"
```

Comment:
nit: for consistency in error wording, does it make sense to align this and the next error?
```suggestion
                error: "Proof of reserve SCI must have the same token id for the required output as for the pseudo-output"
```